File: matlab/tests/test_nested_ma_contraction.m
%% test_nested_ma_contraction.m — multi-attribute nested contraction, and
%  the nested-tuple enumeration used by the Bulger path.
%
%  Mirror of the Python test_nested_ma_contraction.py. Two things are guarded:
%
%  1. MULTI-ATTRIBUTE ROUTING. The fast tree contraction previously handled
%     only a single-attribute density; tensoring any further attribute (e.g.
%     an inversion flag) made it decline. The MA path now routes each nested
%     factor through the contraction and each plain factor through
%     mobius.maPerAttrInnerMatrix, combining them per event-pair (JMM Eq 3.4):
%         <X,Y> = sum_{i,j} prod_a I_a(i,j).
%     Per-attribute prefactors are constant and cancel in the cosine, so
%     mixing the two matrix conventions is exact.
%
%  2. NESTED-TUPLE ENUMERATION (the Bulger / eager build). The leaf
%     enumeration mapped combination positions to slot indices with
%     rowset(nchoosek(1:k, r0)); at r0 = 1 the (k x 1) position column
%     followed the row vector rowset's orientation, collapsing k single-slot
%     combinations into one k-slot combination. Any nesting level with r = 1
%     and k > 1 (e.g. one note read per chord) was mis-enumerated --- a crash
%     in the MA tensor build, wrong tuples in the single-attribute build.
%
%  Reference values are the Python results for identical constructions; the
%  contraction and the (now-correct) Bulger enumeration must agree with them
%  and with each other.

if ~exist('results', 'var')
    results = {};
    standalone = true;
    addpath(fileparts(mfilename('fullpath')));
    clear cleanupDefaults
    cleanupDefaults = mptTestIsolateDefaults(); %#ok<NASGU>
else
    standalone = false;
end

ATOL = 1e-9;     % contract vs bulger within MATLAB
GTOL = 1e-5;     % cross-language vs the Python goldens

C = 0; E = 4; G = 7; Eb = 3;
IVI  = {[C E G], [G 11 2], [C E G]};
ivi  = {[C Eb G], [G 11 2], [C Eb G]};
IVI2 = {[C E G C E G], [G 11 2 G 11 2], [C E G C E G]};
mel  = {[2 5 9], [9 1 4], [2 5 9]};

% --- single-attribute nested: contract == bulger == Python, both arities.
%     The r=1 case is the direct guard for the leaf enumeration fix. ---
ref_sa = containers.Map({1, 2}, {0.667053, 0.111164});
Xsa = {[C E G], [G 11 2], [C E G]};
Ysa = {[C Eb G], [G 11 2], [C Eb G]};
for ri = [1 2]
    X = sadens(Xsa, ri);
    Y = sadens(Ysa, ri);
    cC = cosSimExpTens(X, Y, 'method', 'contract', 'verbose', false);
    cB = cosSimExpTens(X, Y, 'method', 'bulger',   'verbose', false);
    results{end+1, 1} = sprintf('nested-ma: SA ri=%d contract==bulger==Python', ri); %#ok<*SAGROW>
    results{end, 2}   = abs(cC - cB) < ATOL && abs(cC - ref_sa(ri)) < GTOL;
end

% --- MA (nested + flag): flags match (-> harmonic match) and differ (~0) ---
for ri = [1 2]
    X = madens({IVI}, 0.5, ri);
    [cMc, cMb] = bothMethods(X, madens({IVI}, 0.5, ri));
    [cDc, cDb] = bothMethods(X, madens({ivi}, -0.5, ri));
    results{end+1, 1} = sprintf('nested-ma: 1ev flags match ri=%d (=1)', ri);
    results{end, 2}   = abs(cMc - 1) < GTOL && abs(cMb - 1) < GTOL && abs(cMc - cMb) < ATOL;
    results{end+1, 1} = sprintf('nested-ma: 1ev flags differ ri=%d (~0)', ri);
    results{end, 2}   = abs(cDc) < 1e-4 && abs(cDb) < 1e-4 && abs(cDc - cDb) < ATOL;
end

% --- MA, unequal nested cardinality (3 vs 6) plus the flag ---
ref_uneq = containers.Map({1, 2}, {1.0, 0.837924});
for ri = [1 2]
    [cC, cB] = bothMethods(madens({IVI}, 0.5, ri), madens({IVI2}, 0.5, ri));
    results{end+1, 1} = sprintf('nested-ma: 1ev UNEQ 3v6 ri=%d', ri);
    results{end, 2}   = abs(cC - ref_uneq(ri)) < GTOL && abs(cB - ref_uneq(ri)) < GTOL ...
                     && abs(cC - cB) < ATOL;
end

% --- MA, multi-event (2 events per side): per-event-pair combination ---
ref_multi = containers.Map({1, 2}, {0.645572, 0.527059});
for ri = [1 2]
    X = madens({IVI, ivi}, [0.5 -0.5], ri);
    Y = madens({IVI, mel}, [0.5 0.5], ri);
    [cC, cB] = bothMethods(X, Y);
    cSelf = cosSimExpTens(X, X, 'method', 'contract', 'verbose', false);
    results{end+1, 1} = sprintf('nested-ma: 2ev multi-event ri=%d, self=1', ri);
    results{end, 2}   = abs(cC - ref_multi(ri)) < GTOL && abs(cB - ref_multi(ri)) < GTOL ...
                     && abs(cC - cB) < ATOL && abs(cSelf - 1) < GTOL;
end

% --- one-sided normalisation must route through the contraction too ---
%     Regression: before the fix, method='contract' + 'oneSidedDenom' errored
%     and 'auto' fell back to the joint-tuple enumeration. The contraction
%     returns the bare (xy, xx, yy) triple and the denominator is chosen
%     downstream, so it is correct for either normalisation. Doubled chords
%     (IVI2) against the single voicing (IVI), flags matching, give a clean
%     magnitude golden (Python: 8 at rIn=1, 64 at rIn=2; SA and MA alike).
ref_os = containers.Map({1, 2}, {8.0, 64.0});
for ri = [1 2]
    g = ref_os(ri); aTol = ATOL * max(1, g); gTol = GTOL * max(1, g);
    % single nested attribute (internal.nestedContract)
    Xs = sadens(IVI2, ri); Ys = sadens(IVI, ri);
    osC = cosSimExpTens(Xs, Ys, 'method', 'contract', 'normalize', 'oneSidedDenom', 'verbose', false);
    osB = cosSimExpTens(Xs, Ys, 'method', 'bulger',   'normalize', 'oneSidedDenom', 'verbose', false);
    osA = cosSimExpTens(Xs, Ys, 'normalize', 'oneSidedDenom', 'verbose', false);
    results{end+1, 1} = sprintf('nested-ma: SA one-sided ri=%d contract==bulger==auto==Python', ri);
    results{end, 2}   = abs(osC - osB) < aTol && abs(osA - osB) < aTol && abs(osB - g) < gTol;
    % nested (x) flag (internal.nestedContractMA)
    Xm = madens({IVI2}, 0.5, ri); Ym = madens({IVI}, 0.5, ri);
    omC = cosSimExpTens(Xm, Ym, 'method', 'contract', 'normalize', 'oneSidedDenom', 'verbose', false);
    omB = cosSimExpTens(Xm, Ym, 'method', 'bulger',   'normalize', 'oneSidedDenom', 'verbose', false);
    omA = cosSimExpTens(Xm, Ym, 'normalize', 'oneSidedDenom', 'verbose', false);
    results{end+1, 1} = sprintf('nested-ma: MA one-sided ri=%d contract==bulger==auto==Python', ri);
    results{end, 2}   = abs(omC - omB) < aTol && abs(omA - omB) < aTol && abs(omB - g) < gTol;
end


if standalone
    nPass = sum([results{:, 2}]);
    nFail = size(results, 1) - nPass;
    for i = 1:size(results, 1)
        if results{i, 2}
            fprintf('  PASS  %s\n', results{i, 1});
        else
            fprintf('  FAIL  %s\n', results{i, 1});
        end
    end
    fprintf('\n=== test_nested_ma_contraction: %d passed, %d failed (of %d) ===\n\n', ...
        nPass, nFail, nPass + nFail);
    clear cleanupDefaults
    if nFail > 0
        error('test_nested_ma_contraction:failed', '%d test(s) failed.', nFail);
    end
end


% ----------------------------------------------------------------------
function [cC, cB] = bothMethods(X, Y)
    cC = cosSimExpTens(X, Y, 'method', 'contract', 'verbose', false);
    cB = cosSimExpTens(X, Y, 'method', 'bulger',   'verbose', false);
end


function d = sadens(chords, rIn)
    % Single nested harmonic attribute (outer relative, periodic).
    SIG = 0.15; P = 12.0;
    nCh = numel(chords); nValue = numel(chords{1});
    tags = [];
    for k = 1:nCh; tags = [tags, (k - 1) * ones(1, nValue)]; end
    p0 = [];
    for c = 1:nCh; p0 = [p0, chords{c}]; end
    sp = struct('tags', tags, 'r', [rIn nCh], 'sym', [true false], 'rel', [0 1]);
    d = buildExpTens({p0(:)}, {[]}, 'specs', {sp}, 'sigma', SIG, ...
                     'isPer', true, 'period', P, 'verbose', false);
end


function d = madens(events, flags, rIn)
    % Nested harmonic attribute (outer relative, periodic) tensored with a
    % 1-D non-periodic flag attribute. events is a cell over events, each a
    % cell of chords (pitch-class row vectors).
    SIG = 0.15; SF = 0.1; P = 12.0;
    nCh = numel(events{1}); nValue = numel(events{1}{1});
    tags = [];
    for k = 1:nCh; tags = [tags, (k - 1) * ones(1, nValue)]; end
    N = numel(events);
    p0 = zeros(nCh * nValue, N);
    for e = 1:N
        col = [];
        for c = 1:nCh; col = [col, events{e}{c}]; end
        p0(:, e) = col(:);
    end
    p1 = reshape(flags, 1, []);
    sp0 = struct('tags', tags, 'r', [rIn nCh], 'sym', [true false], 'rel', [0 1]);
    sp1 = struct('r', 1, 'sym', false, 'rel', false);
    d = buildExpTens({p0, p1}, {[], []}, 'specs', {sp0, sp1}, ...
                     'sigma', [SIG SF], 'isPer', [true false], ...
                     'period', [P 1.0], 'verbose', false);
end

File: python/tests/test_nested_ma_contraction.py
"""Multi-attribute nested contraction: a nested attribute tensored with one
or more plain attributes.

Previously the fast tree-contraction handled only a single-attribute density;
tensoring any further attribute (e.g. an inversion flag) made it decline, and
the dispatcher fell back to the joint-tuple enumeration --- which at inner
``r = 2`` with wide chords is the combinatorial blow-up the contraction was
built to avoid (minutes per cosine). The MA path now routes each nested factor
through the contraction and each plain factor through the per-attribute MA
matrix, combining them per event-pair (JMM Eq 3.4): ``<X,Y> = Σ_{i,j} Π_a
I_a(i,j)``. Per-attribute prefactors are constant and cancel in the cosine, so
mixing the two matrix conventions is exact.

These check the MA contraction against the exact ``bulger`` enumeration across
single- and multi-event densities, matching and mismatching flags, equal and
unequal nested cardinalities, and both inner arities.
"""
import numpy as np
import pytest

from mpt import build_exp_tens, cos_sim_exp_tens

SIG = 0.15
SF = 0.1          # flag kernel width (small -> strict categorical match)
P = 12.0


def _dens(events, flags, r_in, rel_out=1):
    """Two-attribute density: a nested harmonic attribute (outer relative,
    periodic) tensored with a 1-D non-periodic flag attribute.

    ``events`` is a list over events, each a list of chords (lists of pitch
    classes); ``flags`` a matching list of scalar flag coordinates.
    """
    n_chords = len(events[0])
    n_slots = len(events[0][0])
    tags = np.concatenate([np.full(n_slots, k) for k in range(n_chords)])
    cols = [np.array([v for ch in ev for v in ch], float) for ev in events]
    p_harm = np.stack(cols, axis=1)                       # (slots, N_events)
    p_flag = np.array(flags, float).reshape(1, -1)        # (1, N_events)
    specs = [{"tags": tags, "r": [r_in, n_chords], "sym": [True, False],
              "rel": [0, rel_out]},
             {"r": 1, "sym": False, "rel": False}]
    return build_exp_tens([p_harm, p_flag], None, specs=specs,
                          sigma=[SIG, SF], is_per=[True, False],
                          period=[P, 1.0], verbose=False)


def _both(X, Y):
    c = cos_sim_exp_tens(X, Y, method="contract", verbose=False)
    b = cos_sim_exp_tens(X, Y, method="bulger", verbose=False)
    return c, b


C, E, G, Eb = 0.0, 4.0, 7.0, 3.0
_IVI = [[C, E, G], [G, 11.0, 2.0], [C, E, G]]          # I - V - I (major)
_ivi = [[C, Eb, G], [G, 11.0, 2.0], [C, Eb, G]]        # i - V - i (minor)
_IVI2 = [[C, E, G, C, E, G], [G, 11.0, 2.0, G, 11.0, 2.0],
         [C, E, G, C, E, G]]                            # each chord doubled


@pytest.mark.parametrize("r_in", [1, 2])
def test_ma_nested_flag_match_and_differ(r_in):
    # Flags agree -> joint equals the harmonic match; flags differ -> the
    # strict flag kernel drives the joint to ~0. Both must equal bulger.
    X = _dens([_IVI], [0.5], r_in)
    c, b = _both(X, _dens([_IVI], [0.5], r_in))
    assert c == pytest.approx(b, abs=1e-9) and c == pytest.approx(1.0, abs=1e-9)
    c, b = _both(X, _dens([_ivi], [-0.5], r_in))
    assert c == pytest.approx(b, abs=1e-9) and abs(c) < 1e-6


@pytest.mark.parametrize("r_in", [1, 2])
def test_ma_nested_unequal_cardinality_with_flag(r_in):
    # 3-pitch chords vs the same chords doubled to 6 pitches, plus the flag.
    X = _dens([_IVI], [0.5], r_in)
    Y = _dens([_IVI2], [0.5], r_in)
    c, b = _both(X, Y)
    assert c == pytest.approx(b, abs=1e-9)
    assert c > 1e-3                       # nonzero (the bug returned 0 at r=2)
    assert _both(Y, X)[0] == pytest.approx(c, abs=1e-9)   # symmetric


@pytest.mark.parametrize("r_in", [1, 2])
def test_ma_nested_multi_event(r_in):
    # Two events per side: exercises the per-event-pair matrix combination,
    # not just the single-event factorisation.
    X = _dens([_IVI, _ivi], [0.5, -0.5], r_in)
    Y = _dens([_IVI, [[2., 5., 9.], [9., 1., 4.], [2., 5., 9.]]],
              [0.5, 0.5], r_in)
    c, b = _both(X, Y)
    assert c == pytest.approx(b, abs=1e-9)
    assert _both(X, X)[0] == pytest.approx(1.0, abs=1e-9)


def test_ma_nested_factorises_into_harmonic_times_flag():
    # For single-event densities the cosine factorises exactly into the
    # harmonic cosine (1-attribute contraction) times the flag cosine.
    import math
    r_in = 2
    Xj = _dens([_IVI], [0.5], r_in)
    Yj = _dens([_ivi], [0.5], r_in)         # flags agree -> flag cosine = 1
    joint = cos_sim_exp_tens(Xj, Yj, method="contract", verbose=False)

    def _harm(ev):
        n_slots = len(ev[0])
        tags = np.concatenate([np.full(n_slots, k) for k in range(len(ev))])
        p = np.array([v for ch in ev for v in ch], float).reshape(-1, 1)
        return build_exp_tens([p], None,
                              specs=[{"tags": tags, "r": [r_in, len(ev)],
                                      "sym": [True, False], "rel": [0, 1]}],
                              sigma=[SIG], is_per=[True], period=[P],
                              verbose=False)
    s_harm = cos_sim_exp_tens(_harm(_IVI), _harm(_ivi), method="contract",
                              verbose=False)
    s_flag = math.exp(-(0.5 - 0.5) ** 2 / (4 * SF ** 2))   # = 1
    assert joint == pytest.approx(s_harm * s_flag, abs=1e-9)


# ---------------------------------------------------------------------
# Regression: the nested contraction must also serve one-sided
# normalisation, not cosine alone. Before the fix both gates declined any
# non-cosine normalisation, so method='contract' + 'oneSidedDenom' raised and
# method='auto' fell back to the joint-tuple enumeration (the combinatorial
# blow-up the contraction exists to avoid). The contraction returns the bare
# (xy, xx, yy) triple and the denominator is chosen downstream, so it is
# correct for either normalisation.
# ---------------------------------------------------------------------


def _dens_sa(events, r_in, rel_out=1):
    """Single nested harmonic attribute (no flag): the single-attribute gate."""
    n_chords = len(events[0])
    n_slots = len(events[0][0])
    tags = np.concatenate([np.full(n_slots, k) for k in range(n_chords)])
    cols = [np.array([v for ch in ev for v in ch], float) for ev in events]
    p_harm = np.stack(cols, axis=1)
    specs = [{"tags": tags, "r": [r_in, n_chords], "sym": [True, False],
              "rel": [0, rel_out]}]
    return build_exp_tens([p_harm], None, specs=specs, sigma=[SIG],
                          is_per=[True], period=[P], verbose=False)


@pytest.mark.parametrize("r_in", [1, 2])
def test_one_sided_contraction_matches_bulger_sa(r_in):
    # Single nested attribute. method='contract' previously raised for
    # 'oneSidedDenom'; it must now run and equal the exact enumeration, and
    # the auto dispatch must pick the contraction and agree.
    X = _dens_sa([_IVI2], r_in)        # doubled chords: the costly enumeration case
    Y = _dens_sa([_IVI], r_in)
    contract = cos_sim_exp_tens(X, Y, method="contract",
                                normalize="oneSidedDenom", verbose=False)
    bulger = cos_sim_exp_tens(X, Y, method="bulger",
                              normalize="oneSidedDenom", verbose=False)
    auto = cos_sim_exp_tens(X, Y, normalize="oneSidedDenom", verbose=False)
    assert contract == pytest.approx(bulger, abs=1e-9)
    assert auto == pytest.approx(bulger, abs=1e-9)
    assert np.isfinite(bulger) and bulger > 0.0


@pytest.mark.parametrize("r_in", [1, 2])
def test_one_sided_contraction_matches_bulger_ma(r_in):
    # Multi-attribute (nested (x) flag): the second gate.
    X = _dens([_IVI2], [0.5], r_in)
    Y = _dens([_IVI], [0.5], r_in)
    contract = cos_sim_exp_tens(X, Y, method="contract",
                                normalize="oneSidedDenom", verbose=False)
    bulger = cos_sim_exp_tens(X, Y, method="bulger",
                              normalize="oneSidedDenom", verbose=False)
    auto = cos_sim_exp_tens(X, Y, normalize="oneSidedDenom", verbose=False)
    assert contract == pytest.approx(bulger, abs=1e-9)
    assert auto == pytest.approx(bulger, abs=1e-9)
    assert np.isfinite(bulger) and bulger > 0.0
